open_spiel/games/crazy_eights.cc

```diff
@@ -0,0 +1,657 @@
+    return;
+  }
+
+  if(history_.size() == num_players_ * num_initial_cards_) {
+    SPIEL_CHECK_EQ(current_player_, dealer_);
+    redraw_ = true;
+    return;
+  }
+
+  if(!num_cards_left_) can_pass_action_ = true;
+
+  // if has accumlated 2s and has decided to draw these 2s from previous plays
+  if(start_draw_twos_) {
+    SPIEL_CHECK_TRUE(use_special_cards_);
+    num_draws_from_twos_left_--;
+    // assume if there is no card in the pile then the liability is cleared
+    if(!num_cards_left_) {
+      // if it is due to that the pile is exhausted during drawing +2s, counted as a pass
+      if(!num_draws_from_twos_left_) num_passes_++;
+      num_draws_from_twos_left_ = 0;
+    }
+    if(!num_draws_from_twos_left_) {
+      start_draw_twos_ = false;
+      phase_ = Phase::kPlay;
+      current_player_ = (current_player_ + direction_ + num_players_) % num_players_;
+    }
+    return;
+  }
+
+  // lastly, consider when the player draws card without having a previous +2 card
+  num_draws_before_play_++;
+  phase_ = Phase::kPlay;
+
+  if(!num_cards_left_) num_draws_before_play_ = max_draw_cards_;
+  if(num_draws_before_play_ == max_draw_cards_){
+    can_pass_action_ = true;
+  }
+}
+
+
+
+void SearchLegalCards(std::vector<Action>* legal_actions, 
+    const std::vector<int>& hand, int last_rank, int last_suit){
+  
+  for(int card = 0; card < kNumCards; ++card){
+    if(hand[card] == 0) continue;
+    Suit suit = GetSuit(card);
+    int rank = GetRank(card);
+    if(rank == kEightRank) legal_actions->push_back(card);
+    else if(last_suit == suit || last_rank == rank) legal_actions->push_back(card);
+  }
+  return;
+}
+
+
+
+
+std::vector<Action> CrazyEightsState::PlayLegalActions() const {
+  std::vector<Action> legal_actions;
+  if(nominate_suits_){
+    for(int suit = kClubs; suit <= kSpades; ++suit){
+      legal_actions.push_back(suit+kNominateSuitActionBase);
+    }
+    return legal_actions;
+  }
+
+  if(can_pass_action_ || !num_cards_left_) {
+    SPIEL_CHECK_TRUE(!start_draw_twos_);
+    legal_actions.push_back(kPass);
+  }
+
+  if(num_draws_from_twos_left_){
+    SPIEL_CHECK_GT(num_cards_left_, 0);
+
+  
+    legal_actions.push_back(kDraw);
+    // since we are able to draw
+    SPIEL_CHECK_FALSE(can_pass_action_);
+    SPIEL_CHECK_TRUE(use_special_cards_);
+  
+    if(!start_draw_twos_){
+      for(int suit = kClubs; suit <= kSpades; ++suit){
+        int duo_card = GetAction(static_cast<Suit>(suit), kDrawTwoRank);
+        if(hands_[current_player_][duo_card]) legal_actions.push_back(duo_card);
+      }
+    }
+  } else{
+    SearchLegalCards(&legal_actions, hands_[current_player_], GetRank(last_card_), last_suit_);
+    if(num_cards_left_ && num_draws_before_play_ != max_draw_cards_) {
+      SPIEL_CHECK_FALSE(can_pass_action_);
+      legal_actions.push_back(kDraw);
+    }
+  }
+  absl::c_sort(legal_actions);
+  return legal_actions;
+}
+
+
+bool CrazyEightsState::AfterPlayCard(int action){
```

Comment:
The only purpose of this function is to check if all the cards have been played, right? If so, then maybe a more descriptive name is `CheckAllCardsPlayed`.